File: app/src/main/java/com/apps/game/GameMetadataFormatter.java
package com.apps.game;

import android.text.TextUtils;

import com.core.model.EngineType;
import com.core.model.Game;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 游戏元数据格式化与解析的纯静态工具集。
 *
 * 来源：LauncherLibraryFragment / PadManageFragment / LauncherGameActionController 三处重复实现。
 * 全部为无状态纯函数，可被任意 Fragment / Controller / Adapter 调用。
 */
public final class GameMetadataFormatter {

    private GameMetadataFormatter() {
    }

    /** 返回可读的游玩状态文本（未玩/在玩/玩过），null 或未知值统一返回"未玩"。 */
    public static String playStatusText(String status) {
        if (status == null) return "未玩";
        switch (status) {
            case "playing":
                return "在玩";
            case "completed":
                return "玩过";
            default:
                return "未玩";
        }
    }

    /** 返回可读的引擎名称；null 或未知值返回"未知"。 */
    public static String engineText(EngineType engine) {
        if (engine == null) return "未知";
        switch (engine) {
            case KIRIKIRI:
                return "Kirikiri";
            case ONS:
                return "ONS";
            case TYRANO:
                return "Tyrano";
            case ARTEMIS:
                return "Artemis";
            case WINLATOR:
                return "Winlator";
            case GAMEHUB:
                return "GameHub";
            case PSP:
                return "PSP";
            case NINTENDO_3DS:
                return "3DS";
            default:
                return "未知";
        }
    }

    /**
     * 解析时长字符串为分钟数。
     * 支持 d/h/m/s 单位组合（如 "3h 20m"），纯数字视为分钟。
     * 解析失败或为空时返回 null。
     */
    public static Long parseDuration(String text) {
        if (text == null || text.trim().isEmpty()) return null;
        text = text.trim().toLowerCase(Locale.ROOT);
        try {
            if (!text.matches(".*[dhms].*")) return (long) Double.parseDouble(text);
            long total = 0;
            Matcher m = Pattern.compile("(\\d+(?:\\.\\d+)?)\\s*([dhms])").matcher(text);
            boolean found = false;
            while (m.find()) {
                found = true;
                double val = Double.parseDouble(m.group(1));
                String unit = m.group(2);
                if (unit.equals("d")) total += val * 1440;
                else if (unit.equals("h")) total += val * 60;
                else if (unit.equals("m")) total += val;
                else if (unit.equals("s")) total += val / 60;
            }
            return found ? total : null;
        } catch (Throwable t) {
            return null;
        }
    }

    /**
     * 归一化游玩状态字符串为服务端可识别的三种值：playing / completed / unplayed。
     * null 或空串归一化为 unplayed。
     */
    public static String normalizePlayStatus(String status) {
        String value = status == null ? "" : status.trim().toLowerCase(Locale.ROOT);
        if ("completed".equals(value) || "playing".equals(value)) return value;
        return "unplayed";
    }

    /**
     * 解析开发商字段为列表。null / 空 / "-" 返回空列表。
     * 按 / 、 , ， 切分，去重并 trim。
     */
    public static List<String> parseDevelopers(String developersText) {
        List<String> result = new ArrayList<>();
        if (developersText == null || developersText.trim().isEmpty()
                || "-".equals(developersText.trim())) return result;
        String[] parts = developersText.split("/|、|,|，");
        for (String raw : parts) {
            String developer = raw == null ? "" : raw.trim();
            if (!developer.isEmpty() && !result.contains(developer)) result.add(developer);
        }
        return result;
    }

    /** 返回游戏标题；空或缺失时返回"未命名游戏"。 */
    public static String safeTitle(Game game) {
        if (game == null || TextUtils.isEmpty(game.title)) return "未命名游戏";
        return game.title.trim();
    }
}

File: app/src/main/java/com/apps/game/GameMetadataFormatter.kt
package com.apps.game

import android.content.Context
import android.text.TextUtils
import com.core.R
import com.core.model.EngineType
import com.core.model.Game
import java.util.Locale
import java.util.regex.Pattern

/**
 * 游戏元数据格式化与解析的纯静态工具集。
 *
 * 来源：LauncherLibraryFragment / PadManageFragment / PadGameFragment 三处重复实现。
 * 全部为无状态纯函数，可被任意 Fragment / Controller / Adapter 调用。
 */
object GameMetadataFormatter {

    /** 返回可读的游玩状态文本（未玩/在玩/玩过），null 或未知值统一返回"未玩"。 */
    @JvmStatic
    fun playStatusText(status: String?): String = when (status) {
        "playing" -> "Playing"
        "completed" -> "Completed"
        else -> "Not started"
    }

    @JvmStatic
    fun playStatusText(context: Context, status: String?): String = context.getString(
        when (status) {
            "playing" -> R.string.game_status_playing
            "completed" -> R.string.game_status_completed
            else -> R.string.game_status_unplayed
        }
    )

    /** 返回可读的引擎名称；null 或未知值返回"未知"。 */
    @JvmStatic
    fun engineText(context: Context, engine: EngineType?): String = context.getString(
        when (engine) {
            EngineType.KIRIKIRI -> R.string.game_engine_kirikiri
            EngineType.ONS -> R.string.game_engine_onscripter
            EngineType.TYRANO -> R.string.game_engine_tyrano
            EngineType.ARTEMIS -> R.string.game_engine_artemis
            EngineType.WINLATOR -> R.string.game_engine_winlator
            EngineType.GAMEHUB -> R.string.game_engine_gamehub
            EngineType.PSP -> R.string.game_engine_psp
            EngineType.NINTENDO_3DS -> R.string.game_engine_nintendo_3ds
            EngineType.NINTENDO_SWITCH -> R.string.game_engine_nintendo_switch
            else -> R.string.game_common_unknown
        }
    )

    /**
     * 解析时长字符串为分钟数。
     * 支持 d/h/m/s 单位组合（如 "3h 20m"），纯数字视为分钟。
     * 解析失败或为空时返回 null。
     */
    @JvmStatic
    fun parseDuration(text: String?): Long? {
        if (text == null || text.trim { it <= ' ' }.isEmpty()) return null
        val normalized = text.trim { it <= ' ' }.lowercase(Locale.ROOT)
        return try {
            if (!normalized.matches(".*[dhms].*".toRegex())) {
                return normalized.toDouble().toLong()
            }
            var total = 0L
            val m = Pattern.compile("(\\d+(?:\\.\\d+)?)\\s*([dhms])").matcher(normalized)
            var found = false
            while (m.find()) {
                found = true
                val v = m.group(1)!!.toDouble()
                val delta = when (m.group(2)) {
                    "d" -> v * 1440
                    "h" -> v * 60
                    "m" -> v
                    "s" -> v / 60
                    else -> 0.0
                }
                total = (total + delta).toLong()
            }
            if (found) total else null
        } catch (error: RuntimeException) {
            // 时长字符串解析失败时返回 null，由调用方按无数据处理（契约见 KDoc），可安全忽略
            null
        }
    }

    /**
     * 归一化游玩状态字符串为服务端可识别的三种值：playing / completed / unplayed。
     * null 或空串归一化为 unplayed。
     */
    @JvmStatic
    fun normalizePlayStatus(status: String?): String {
        val value = status?.trim { it <= ' ' }?.lowercase(Locale.ROOT) ?: ""
        return if (value == "completed" || value == "playing") value else "unplayed"
    }

    /**
     * 解析开发商字段为列表。null / 空 / "-" 返回空列表。
     * 按 / 、 , ， 切分，去重并 trim。
     */
    @JvmStatic
    fun parseDevelopers(developersText: String?): List<String> {
        if (developersText == null || developersText.trim { it <= ' ' }.isEmpty()
            || "-" == developersText.trim { it <= ' ' }
        ) {
            return emptyList()
        }
        return developersText.split("/|、|,|，".toRegex())
            .map { it.trim { it <= ' ' } }
            .filter { it.isNotEmpty() }
            .distinct()
    }

    /** 返回游戏标题；空或缺失时返回"未命名游戏"。 */
    @JvmStatic
    fun safeTitle(game: Game?): String {
        if (game == null || TextUtils.isEmpty(game.title)) return "Game"
        return game.title!!.trim { it <= ' ' }
    }

    @JvmStatic
    fun safeTitle(context: Context, game: Game?): String {
        if (game == null || TextUtils.isEmpty(game.title)) {
            return context.getString(R.string.game_unnamed)
        }
        return game.title!!.trim { it <= ' ' }
    }
}
